### ISSUE TITLE
[RFC]: AscendC categorical sampling backend for NPU Model Runner V2

### ISSUE THREAD

**Issue by freyfwt** [vllm-project/vllm-ascend#14130]
### Motivation.

NPU Model Runner V2 (MRV2) uses categorical sampling on the decode critical path and also relies on the same sampling primitive to resample rejected or bonus tokens during speculative decoding. The vLLM v0.26 sampler contract has evolved beyond the categorical path currently available on Ascend: it supports optional in-kernel temperature application, processed-logits cache writes with scalar or per-token cache columns, a higher-precision `use_fp64` mode, zero-row-stride broadcast logits, and explicit contiguity normalization for selected metadata tensors.

Keeping a partially compatible backend is risky. Ordinary sampling and speculative decoding can call the same helper with different combinations of these arguments, so a missing feature can appear only under a particular proposer, cache layout, rejection outcome, or precision mode. Silent fallback is also undesirable: it can change random-number semantics, introduce graph breaks, or hide a large performance regression on the decode path.

The existing Triton-Ascend implementation is not a complete solution for this contract. In particular, Ascend Triton does not provide the upstream FP64 random and reduction path, and a blockwise Gumbel-max implementation creates intermediate state proportional to the number of vocabulary blocks. A native AscendC operator can instead perform exact categorical selection from stable exponential weights, use row-local stateless RNG, handle large vocabularies without a full-vocabulary random tensor, and keep validation on device during ACLGraph replay.

This is an architectural decision because it introduces a native operator contract, defines failure semantics for invalid device data, selects different custom-op loading paths on A2/A3 and A5, and requires hardware-specific CI coverage. It is related to, but separate from, #9269. That RFC concerns the MRV1 sampler adapter, batch-parallel sampling, and logits-processing modes; this RFC defines the MRV2 categorical backend that MRV1 may reuse later.

#### Summary

Add one AscendC-backed categorical sampling operator for MRV2 and route both the ordinary sampler and speculative rejection resampling through one upstream-compatible Python wrapper. The implementation will preserve the vLLM v0.26 `gumbel_sample` call contract while using categorical inverse-CDF selection internally. It will support A2, A3, and A5 without CPU fallback, host synchronization, or a user-facing feature flag.

#### Goals

- Match the vLLM v0.26 Python sampling signature and observable argument semantics.
- Support greedy and categorical rows, optional temperature application, processed-logits cache writes, zero-row-stride logits, and the higher-precision sampling path.
- Use the same wrapper from the ordinary MRV2 sampler and its speculators, including FP64 rejection resampling.
- Preserve ACLGraph capture and replay without `.item()`, diagnostic output tensors, or other normal-path device-to-host synchronization.
- Provide the same public behavior on A2, A3, and A5, with CI that fails rather than skips when the operator is not built or registered.
- Keep random sampling reproducible from request seed and logical position, independent of unrelated batch rows and graph padding.

#### Non-goals

- This RFC does not add synthetic rejection sampling or block verification to MRV2. Their existing unsupported/fallback behavior is unchanged and should be handled by focused follow-up work.
- This RFC does not change MRV1 routing, batch-parallel sampling, or the logits-processing modes proposed in #9269.
- It does not require token-for-token equality with CUDA, Triton, or a different Ascend generation. Categorical distributions and documented reproducibility boundaries must match; implementation-specific random streams need not.
- It does not fuse top-k, top-p, penalties, grammar masking, or other upstream logits processors. The operator consumes the processed logits produced at the existing sampling boundary.
- It does not expose new user configuration or environment variables.

#### Interface and ownership

The vLLM-facing Python wrapper keeps the v0.26 signature:

```python
gumbel_sample(
    logits,
    expanded_idx_mapping,
    temperature,
    seed,
    pos,
    apply_temperature,
    output_processed_logits=None,
    output_processed_logits_col=None,
    use_fp64=False,
) -> sampled_token_ids
```

To preserve upstream layout semantics, the wrapper makes only `expanded_idx_mapping`, `pos`, and the optional `output_processed_logits_col` contiguous before native dispatch. It must not silently cast or make `logits`, `temperature`, `seed`, or `output_processed_logits` contiguous. Invalid device, dtype, rank, shape, or layout combinations are errors.

The native operator has one stable Torch/ACLNN schema:

```text
npu_categorical_sample(
    processed_logits,
    expanded_idx_mapping,
    temperature,
    seed,
    pos,
    return_lse=False,
    apply_temperature=False,
    output_processed_logits=None,
    output_processed_logits_col=None,
    use_fp64=False,
) -> (sampled_token_ids, lse)
```

`processed_logits` has shape `[N, V]`, dtype FP16, BF16, or FP32, and a contiguous vocabulary dimension. Normal padded row strides and a zero row stride used by broadcast logits are supported; arbitrary overlapping rows are not. The initial production limit is `1 <= V <= 1,048,576`.

`expanded_idx_mapping` maps each logits row to request state. `-1` is an ACLGraph padding row. `temperature`, `seed`, and `pos` retain their existing request-state and row-state meanings. `sampled_token_ids` is `int64[N]`.

When `apply_temperature=True`, a nonzero mapped temperature is applied before sampling and before any processed-logits cache write. Temperature zero remains the greedy sentinel and selects the first maximum token independently of the seed.

`output_processed_logits` is an optional FP32 request-indexed cache with shape `[R, V]` or `[R, C, V]`. The optional `output_processed_logits_col` selects the cache column using either one scalar `int32` value or one value per logits row. A padding row does not write the cache. Out-of-range request mappings and cache columns must fail before the corresponding metadata or cache address is dereferenced.

The native ABI keeps a fixed pair of outputs for Torch schema and ACLGraph stability. When `return_lse=True`, `lse` is FP32 `[N]`; when false, the public result is a zero-length FP32 tensor and the final logarithm/writeback is skipped. The MRV2 wrapper requests no LSE and returns only sampled tokens. LSE remains available for future logprob consumers without adding full-vocabulary probability materialization.

#### Sampling and precision semantics

For a valid categorical row, the operator computes stable weights without materializing probabilities:

```text
m = max(logits)
w[i] = exp(logits[i] - m)
sample the first j where prefix_sum(w, j) > u * sum(w)
```

Greedy rows use the same validation rules but select the first maximum. FP16 and BF16 inputs are converted to FP32 working values. Masked `-inf` entries have zero mass and cannot be selected.

`use_fp64` is an inherited semantic name, not a promise that the NPU stores logits or performs vector math in IEEE FP64. On Ascend it selects a high-precision categorical path using a 64-bit Philox-derived random value and 64-bit fixed-point mass accumulation/selection. Input, cache, and output dtypes do not change. This path exists to preserve meaningful tail resolution and rejection-resampling accuracy where the default FP32 random/reduction path is insufficient; it is not enabled by default.

Randomness is stateless and row-local:

```text
request_state = expanded_idx_mapping[row]
key = seed[request_state]
counter = pos[row]
```

Greedy and padding rows consume no mutable RNG state. Identical inputs on the same supported software/hardware stack are reproducible. Batch reordering, unrelated requests, and inserted padding rows must not change a request's result.

#### Exceptional values and failure semantics

Validation is split by ownership:

1. The Torch binding checks device, dtype, rank, shape, contiguity, and stride constraints known at dispatch.
2. Host tiling independently rejects invalid static descriptions or unavailable platform capability.
3. AscendC checks data-dependent conditions using a static diagnostic followed by a device trap, in both eager and ACLGraph execution.

The row contract is:

| Row condition | Categorical | Greedy | LSE |
| --- | --- | --- | --- |
| Finite values with optional `-inf` masks | Inverse-CDF sample | First maximum | Normal FP32 LSE |
| Contains NaN | Device assertion | Device assertion | No valid output |
| Contains `k` positive infinities and no NaN | Uniform among the `k` positions | First positive infinity | `+inf` |
| Contains only `-inf` | Device assertion | Device assertion | No valid output |
| Mapping is `-1` | Token `0` padding result; no RNG/cache access | Same | `0` if requested |
| Invalid request mapping or cache column | Device assertion before unsafe access | Same | No valid output |

Validation must precede externally visible row writes. Positive infinity is accepted mathematical behavior and produces no warning. Processed logprobs for positive-infinity rows are outside this contract because `+inf - +inf` is undefined without also exposing support-set size. The operator will not return `pos_inf_count`, `row_flags`, or other diagnostic tensors; that would add hot-path memory traffic and require a host-side consumer or synchronization.

#### Integration and hardware routing

The ordinary MRV2 sampler receives already processed logits and calls the wrapper with `apply_temperature=False`. Speculators use the same wrapper, preserving upstream temperature and processed-logits-cache semantics. When speculative rejection sampling requests `use_fp64=True`, the rejection path materializes only the rejected/bonus residual distribution per request and delegates the final categorical draw to this operator.

A2 and A3 use the general custom-op initialization path. A5 loads the ACLNN operator from its supported custom OPP package and registers the same Torch schema without globally enabling unrelated A5 custom operators. Operator absence or registration failure is a clear error; the production call must not silently fall back to CPU or a different sampling algorithm.

The normal path must be ACLGraph-capturable and must not read device values on the host. Host tiling uses shapes, strides, attributes, and platform information only; greedy-versus-categorical selection remains device data so mixed batches can replay without recapture.

#### Compatibility, rollout, and rollback

This change adds no user-facing API or configuration. It aligns the Ascend MRV2 backend with the existing vLLM v0.26 sampler contract. Existing request-level seed/position reproducibility is preserved within the documented backend boundary, but exact sampled tokens may differ from the former Triton Gumbel implementation because the random mapping and categorical algorithm differ while representing the same target distribution.

Rollout should be staged as reviewable native-operator, MRV2 integration, and CI/test changes if maintainers prefer multiple PRs. The wrapper becomes the single production path only after the acceptance gates below pass. The old Triton implementation should not remain as a hidden runtime fallback; rollback is performed by reverting production routing or the integrating PR, making behavior explicit and reviewable.

#### Risks and alternatives

- **Cross-generation CANN compatibility:** assertion initialization and custom-op loading differ across A2/A3/A5. Mitigation: keep hardware-specific bootstrap localized while sharing one schema and kernel contract, and compile plus execute the operator in CI on every generation.
- **High-precision path differs from CUDA FP64 implementation:** fixed-point accumulation is not bit-identical to upstream FP64 Gumbel-max. Mitigation: define distribution, heavy-tail resolution, and reproducibility properties rather than cross-backend token identity, and validate against an FP64 CPU reference.
- **Device assertion behavior can vary by launch mode:** a compile-only check is insufficient. Mitigation: exercise each data-dependent failure in eager and ACLGraph modes and require an observable diagnostic failure.
- **FP64 rejection residual memory:** materializing `[num_requests, vocab_size]` FP32 residual logits can be large. Mitigation: allocate it only when high-precision rejection resampling is requested, split oversized launches, and cover the largest supported production shape in memory and performance validation.
- **Native operator maintenance cost:** a custom kernel adds CANN-facing code. Mitigation: keep the Python interface identical to upstream, isolate the native implementation, and protect its contract with upstream-derived tests so future main-to-main updates fail visibly.

Alternatives considered:

- **Keep the existing Triton Gumbel path:** rejected because it cannot provide the required high-precision path on Ascend and retains backend-specific intermediate and grid limitations.
- **Generate a full `[N, V]` random tensor:** rejected because it adds output-sized temporary memory and memory bandwidth on every decode step.
- **Return row-status tensors and log warnings on the host:** rejected because it adds output traffic, synchronization/consumption complexity, and delayed failure for invalid memory indices.
- **Silently fall back when the custom op is unavailable:** rejected because it hides packaging failures and changes graph, performance, and RNG semantics.

#### Test strategy and acceptance criteria

Acceptance requires all of the following:

- **Interface parity:** signature, optional arguments, metadata contiguity, cache shapes/columns, zero-row-stride logits, padding rows, and static validation are covered against the vLLM v0.26 contract.
- **Operator correctness on A2, A3, and A5:** each generation builds and registers the real custom operator and runs the same compact mandatory NPU suite for FP16, BF16, and FP32; greedy/categorical rows; temperature on/off; cache off/scalar-column/per-token-column; finite, masked, padding, and positive-infinity rows; and representative aligned/non-aligned vocabularies.
- **No false CI success:** required hardware jobs fail if the native operator is missing. They must not skip, mock, use a CPU reference as the system under test, or silently execute the old path. A broader boundary matrix may run nightly, but every PR must exercise a real eager and ACLGraph call on A2, A3, and A5.
- **Failure behavior:** NaN, all-`-inf`, invalid request mapping, and invalid cache-column inputs fail observably through the intended device assertion in eager and ACLGraph modes on all three generations.
- **Reproducibility and distribution:** repeated `(seed, pos)` inputs reproduce; batch permutation and zero-stride broadcast do not alter per-request results; statistical tests against FP64 CPU reference cover uniform, skewed, masked, tied, and heavy-tail distributions at a predeclared significance threshold.
- **Meaningful high-precision validation:** at least one heavy-tail or residual-distribution test must distinguish `use_fp64=True` from merely aliasing the default path. It must validate precision-sensitive probability mass or selection behavior against an FP64 reference, not only confirm that the branch executes.
- **Speculative decoding:** an MRV2 end-to-end or PR-level integration case forces rejection/bonus resampling with `use_fp64=True`, validates accepted/rejected output structure, and verifies that scalar and per-token processed-logits cache columns contain the expected values.
- **Large-shape safety:** batch 128 with vocabulary 1,048,576 completes without grid-limit failure, out-of-bounds access, or unbounded private workspace. Launch splitting is covered wherever a backend grid limit would otherwise be exceeded.
- **ACLGraph:** capture and replay support changing seed/position data and mixed greedy/categorical rows without graph breaks, CPU fallback, or normal-path host synchronization.
- **Performance:** paired same-stack NPU Event measurements compare the new path with the existing production path for representative `B={1, 8, 32, 128}` and `V={32K, 151936, 1M}` cases, including greedy, categorical, cache write, and FP64 rejection-resampling modes where applicable. Any repeatable regression greater than 5% in a production anchor requires an explicit maintainer-approved rationale before merge; averages must not hide a regressed anchor.

Exact test files, fixtures, commands, hardware/software versions, and observed results belong in the implementation PR rather than this RFC.

### Feedback Period.

At least one week, from 2026-08-12 through 2026-08-19.

### CC List.

- @zzzzwwjj for native custom operators and CANN integration.
- @wangxiyuan for speculative decoding and overall Ascend architecture.
- @realliujiaxu and @whx-sjtu for sampling-path review.
- @MengqingCao for model-runner integration.

### Any Other Things.

Related RFC: #9269. This proposal deliberately does not amend the MRV1 configuration or batch-parallel design in that issue. If both proposals are accepted, MRV1 may reuse this categorical backend through the adapter boundary defined by #9269.

A draft implementation may be used as feasibility evidence during the feedback period, but implementation PRs should link this RFC, call out any deviation from the approved contract, and record exact A2/A3/A5 test and performance evidence before merge.
